### PR TITLE
solving the issue 223 by enforcing withdraw instead of postponement

### PR DESCRIPTION
**Pull request by josojo**
solving issue 223 by doing the withdraw instead of postponing them. This change also requires us to make withdraws processable by anyone, especially the solution submitter

---
testplan: unit tests

### CODE REVIEW

File: contracts/EpochTokenLocker.sol [hunk 1]
```diff
@@ -124,9 +128,14 @@ contract EpochTokenLocker {
     /**
      * internal functions
      */
+    function hasValidWithdrawRequest(address user, address token) internal view returns(bool) {
+        return balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentBatchId() &&
+            balanceStates[user][token].pendingWithdraws.stateIndex > 0;
+    }
+
     function addBalanceAndPostponeWithdraw(address user, address token, uint amount) internal {
```
Comment (josojo) [inline contracts/EpochTokenLocker.sol]: Rename function here

File: contracts/EpochTokenLocker.sol [hunk 2]
```diff
@@ -59,28 +59,32 @@ contract EpochTokenLocker {
...
+        if (hasValidWithdrawRequest(msg.sender, token)) {
+            withdraw(token, msg.sender);
+        }
         balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: getCurrentBatchId() });
         emit WithdrawRequest(msg.sender, token, amount, getCurrentBatchId());
     }
 
-    function withdraw(address token) public {
-        updateDepositsBalance(msg.sender, token); // withdrawn amount might just be deposited before
+    function withdraw(address token, address owner) public {
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:70]: Why this change? Can't you keep using `msg.sender` here?

File: contracts/EpochTokenLocker.sol [hunk 3]
```diff
@@ -124,9 +128,14 @@ contract EpochTokenLocker {
     /**
      * internal functions
      */
+    function hasValidWithdrawRequest(address user, address token) internal view returns(bool) {
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol]: Does it maybe make sense to expose this publicly since it's just a view function anyways?

File: contracts/EpochTokenLocker.sol [hunk 4]
```diff
@@ -124,9 +128,14 @@ contract EpochTokenLocker {
...
+        return balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentBatchId() &&
+            balanceStates[user][token].pendingWithdraws.stateIndex > 0;
+    }
+
     function addBalanceAndPostponeWithdraw(address user, address token, uint amount) internal {
-        if (balanceStates[user][token].pendingWithdraws.stateIndex < getCurrentBatchId()) {
-            balanceStates[user][token].pendingWithdraws.stateIndex = getCurrentBatchId();
+        if (hasValidWithdrawRequest(user, token)) {
+            withdraw(token, user);
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:138]: This is a pretty drastic change in logic. Instead of postponing the withdraw we now have the solver execute it. Do we really want this? This means you can offset the gas cost of withdraws to the first solver as long as you keep trading. What's the difference in gas cost for the solver? If I request a bunch of withdraws, can it make a substantial difference to the amount of trades the first solution could have?

I can see an argument for having the same behaviour between `requestWithdraw` and `addBalanceAndPostponeWithdraw` but since the actor calling this method is different, I see some game theoretic problems.

**Review by fleupold — COMMENTED** [review]
referencing #223 

Let's discuss if we also want to execute the withdraw on solution submission.

File: contracts/EpochTokenLocker.sol (hunk 2 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:70]: If the solution submitter is doing the withdraw for another user with an open withdrawalRequest, then I must be enabled for anyone.

File: contracts/EpochTokenLocker.sol (hunk 4 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:138]: Here are the gas costs for calling a withdraw request:
```
··································|·····································|·············|·············|·············|···············|··············
|  EpochTokenLocker               ·  withdraw                           ·      65355  ·      65362  ·      65359  ·            2  ·       0.23  │
··································|·····································|·············|·············|·············|···

```
If we substract the 21000 gas for initiating a transaction, we get 44k gas for processing a withdraw.
That means for 25 orders with open withdraw requests, we would pay an additional 1100k gas. While it is significant, it is acceptable, I guess.

Martin proposed anyways to cap the number of orders to be touched to a certain number, such that all players have the same gas constraints, no matter when and in which order they are submitting their solution. 
Now, this additional gas for the withdraws might just induce a lower upper bound for the number of orders, we can process.

Comment (bh2smith) [reply contracts/EpochTokenLocker.sol:138]: > 1100k

Is this 1.1 Million? If so, then we may get away with this. We determined the following: 
- Average block gas consumption lately has been about 9-10 Million per block. 
- We thought that occupying the entire block would be expensive and difficult to do, so we estimated a reasonable consumption for `submitSolution` to be about 4-4.5 million.
- The relevant estimate from #231  was the ~3.5M for submitting a competing solution with 20 touched orders. 
- This was the reason we went with a max of 25 touched orders in #232, but we could also drop this to a 20 order limit and easily accommodate for these withdraws.

Note, however, that these estimates will likely increase as we have uncovered some issues with the current implementation of the objective function in #229. We will likely have to run a third loop over orders exclusively for accumulating the disregarded utility.

**Comment by josojo** [comment]
Closing, as I would favor PR #233
